File: reports/code-architecture/代码案例/TypeScript案例/三层架构示例.ts
# TypeScript 代码架构案例

## 三层架构示例

### 项目结构
```
src/
├── presentation/
│   └── user.controller.ts
├── business/
│   ├── user.service.ts
│   └── validation.error.ts
├── data/
│   ├── user.model.ts
│   └── user.repository.ts
├── types/
│   └── index.ts
└── main.ts
```

### 完整代码

#### 类型定义
```typescript
// types/index.ts
export interface User {
  id?: number;
  name: string;
  email: string;
  createdAt?: Date;
}

export interface CreateUserDto {
  name: string;
  email: string;
}

export interface UpdateUserDto {
  name?: string;
  email?: string;
}
```

#### 数据层 (Data Layer)
```typescript
// data/user.model.ts
export class UserModel {
  constructor(
    public id: number,
    public name: string,
    public email: string,
    public createdAt: Date = new Date()
  ) {}
}

// data/user.repository.ts
import { User, CreateUserDto, UpdateUserDto } from '../types';
import { UserModel } from './user.model';

interface Database {
  query(sql: string, params?: any[]): any[];
  run(sql: string, params?: any[]): { lastID: number; changes: number };
}

export class UserRepository {
  constructor(private db: Database) {}

  private mapToModel(row: any): UserModel | null {
    if (!row) return null;
    return new UserModel(row.id, row.name, row.email, new Date(row.created_at));
  }

  findById(id: number): UserModel | null {
    const rows = this.db.query('SELECT * FROM users WHERE id = ?', [id]);
    return this.mapToModel(rows[0]);
  }

  findByEmail(email: string): UserModel | null {
    const rows = this.db.query('SELECT * FROM users WHERE email = ?', [email]);
    return this.mapToModel(rows[0]);
  }

  findAll(): UserModel[] {
    const rows = this.db.query('SELECT * FROM users ORDER BY created_at DESC');
    return rows.map(row => this.mapToModel(row)!);
  }

  create(dto: CreateUserDto): UserModel {
    const result = this.db.run(
      'INSERT INTO users (name, email) VALUES (?, ?)',
      [dto.name, dto.email]
    );
    return new UserModel(result.lastID, dto.name, dto.email);
  }

  update(id: number, dto: UpdateUserDto): UserModel | null {
    const user = this.findById(id);
    if (!user) return null;

    const updates: string[] = [];
    const params: any[] = [];

    if (dto.name !== undefined) {
      updates.push('name = ?');
      params.push(dto.name);
    }
    if (dto.email !== undefined) {
      updates.push('email = ?');
      params.push(dto.email);
    }

    if (updates.length === 0) return user;

    params.push(id);
    this.db.run(`UPDATE users SET ${updates.join(', ')} WHERE id = ?`, params);
    
    return this.findById(id);
  }

  delete(id: number): boolean {
    const result = this.db.run('DELETE FROM users WHERE id = ?', [id]);
    return result.changes > 0;
  }
}
```

#### 业务层 (Business Layer)
```typescript
// business/validation.error.ts
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

// business/user.service.ts
import { User, CreateUserDto, UpdateUserDto } from '../types';
import { UserRepository } from '../data/user.repository';
import { ValidationError } from './validation.error';

export class UserService {
  constructor(private userRepo: UserRepository) {}

  private validateName(name: string): void {
    if (!name || name.trim().length < 2) {
      throw new ValidationError('用户名至少2个字符');
    }
    if (name.length > 50) {
      throw new ValidationError('用户名不能超过50个字符');
    }
  }

  private validateEmail(email: string): void {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(email)) {
      throw new ValidationError('邮箱格式不正确');
    }
  }

  create(dto: CreateUserDto): User {
    this.validateName(dto.name);
    this.validateEmail(dto.email);

    // 检查邮箱是否已存在
    const existing = this.userRepo.findByEmail(dto.email);
    if (existing) {
      throw new ValidationError('邮箱已被注册');
    }

    return this.userRepo.create(dto);
  }

  update(id: number, dto: UpdateUserDto): User {
    const user = this.userRepo.findById(id);
    if (!user) {
      throw new ValidationError('用户不存在');
    }

    if (dto.name) {
      this.validateName(dto.name);
    }
    if (dto.email) {
      this.validateEmail(dto.email);
      // 检查新邮箱是否被占用
      const existing = this.userRepo.findByEmail(dto.email);
      if (existing && existing.id !== id) {
        throw new ValidationError('邮箱已被注册');
      }
    }

    const updated = this.userRepo.update(id, dto);
    if (!updated) {
      throw new ValidationError('更新失败');
    }
    return updated;
  }

  delete(id: number): boolean {
    const user = this.userRepo.findById(id);
    if (!user) {
      throw new ValidationError('用户不存在');
    }
    return this.userRepo.delete(id);
  }

  findById(id: number): User | null {
    return this.userRepo.findById(id);
  }

  findAll(): User[] {
    return this.userRepo.findAll();
  }
}
```

#### 表现层 (Presentation Layer)
```typescript
// presentation/user.controller.ts
import { User, CreateUserDto, UpdateUserDto } from '../types';
import { UserService } from '../business/user.service';
import { ValidationError } from '../business/validation.error';

export interface HttpResponse {
  statusCode: number;
  body: string;
}

export class UserController {
  constructor(private userService: UserService) {}

  private success(data: any): HttpResponse {
    return {
      statusCode: 200,
      body: JSON.stringify({ success: true, data })
    };
  }

  private error(message: string, statusCode: number = 400): HttpResponse {
    return {
      statusCode,
      body: JSON.stringify({ success: false, error: message })
    };
  }

  list(): HttpResponse {
    try {
      const users = this.userService.findAll();
      return this.success(users);
    } catch (e) {
      return this.error(e.message);
    }
  }

  get(id: number): HttpResponse {
    try {
      const user = this.userService.findById(id);
      if (!user) {
        return this.error('用户不存在', 404);
      }
      return this.success(user);
    } catch (e) {
      return this.error(e.message);
    }
  }

  create(body: CreateUserDto): HttpResponse {
    try {
      const user = this.userService.create(body);
      return this.success(user);
    } catch (e) {
      if (e instanceof ValidationError) {
        return this.error(e.message);
      }
      return this.error('创建失败');
    }
  }

  update(id: number, body: UpdateUserDto): HttpResponse {
    try {
      const user = this.userService.update(id, body);
      return this.success(user);
    } catch (e) {
      if (e instanceof ValidationError) {
        return this.error(e.message);
      }
      return this.error('更新失败');
    }
  }

  delete(id: number): HttpResponse {
    try {
      const result = this.userService.delete(id);
      return this.success({ deleted: result });
    } catch (e) {
      if (e instanceof ValidationError) {
        return this.error(e.message);
      }
      return this.error('删除失败');
    }
  }
}
```

---

## 依赖注入示例

```typescript
// main.ts
import express from 'express';
import { UserRepository } from './data/user.repository';
import { UserService } from './business/user.service';
import { UserController } from './presentation/user.controller';

// 模拟数据库
const mockDb = {
  query: (sql: string, params: any[]) => [],
  run: (sql: string, params: any[]) => ({ lastID: 1, changes: 1 })
};

// 依赖注入
const userRepo = new UserRepository(mockDb);
const userService = new UserService(userRepo);
const userController = new UserController(userService);

// Express 路由
const app = express();
app.use(express.json());

app.get('/users', (req, res) => {
  const response = userController.list();
  res.status(response.statusCode).send(response.body);
});

app.get('/users/:id', (req, res) => {
  const response = userController.get(parseInt(req.params.id));
  res.status(response.statusCode).send(response.body);
});

app.post('/users', (req, res) => {
  const response = userController.create(req.body);
  res.status(response.statusCode).send(response.body);
});

app.put('/users/:id', (req, res) => {
  const response = userController.update(parseInt(req.params.id), req.body);
  res.status(response.statusCode).send(response.body);
});

app.delete('/users/:id', (req, res) => {
  const response = userController.delete(parseInt(req.params.id));
  res.status(response.statusCode).send(response.body);
});

app.listen(3000, () => {
  console.log('Server running on port 3000');
});
```

---

## NestJS 风格示例

```typescript
// users/user.entity.ts
import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';

@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @Column({ unique: true })
  email: string;

  @Column({ default: () => 'CURRENT_TIMESTAMP' })
  createdAt: Date;
}

// users/dto/create-user.dto.ts
export class CreateUserDto {
  name: string;
  email: string;
}

// users/users.service.ts
import { Injectable, ValidationError } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from './user.entity';
import { CreateUserDto } from './dto/create-user.dto';

@Injectable()
export class UsersService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async create(dto: CreateUserDto): Promise<User> {
    const existing = await this.userRepository.findOne({ where: { email: dto.email } });
    if (existing) {
      throw new ValidationError('邮箱已被注册');
    }
    
    const user = this.userRepository.create(dto);
    return this.userRepository.save(user);
  }

  async findAll(): Promise<User[]> {
    return this.userRepository.find();
  }

  async findOne(id: number): Promise<User> {
    return this.userRepository.findOne({ where: { id } });
  }

  async delete(id: number): Promise<void> {
    await this.userRepository.delete(id);
  }
}

// users/users.controller.ts
import { Controller, Get, Post, Body, Param, Delete } from '@nestjs/common';
import { UsersService } from './users.service';
import { CreateUserDto } from './dto/create-user.dto';

@Controller('users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Post()
  create(@Body() createUserDto: CreateUserDto) {
    return this.usersService.create(createUserDto);
  }

  @Get()
  findAll() {
    return this.usersService.findAll();
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.usersService.findOne(+id);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.usersService.remove(+id);
  }
}
```

File: reports/code-architecture/代码案例/Python案例/三层架构示例.py
# Python 代码架构案例

## 三层架构示例

### 项目结构
```
project/
├── presentation/
│   └── user_view.py
├── business/
│   └── user_service.py
├── data/
│   ├── user_model.py
│   └── user_repository.py
└── main.py
```

### 完整代码

#### 数据层 (Data Layer)
```python
# data/user_model.py
from dataclasses import dataclass
from datetime import datetime

@dataclass
class User:
    id: int = None
    name: str = ""
    email: str = ""
    created_at: datetime = None
    
    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.now()

# data/user_repository.py
from data.user_model import User

class UserRepository:
    def __init__(self, db_connection):
        self.db = db_connection
    
    def find_by_id(self, user_id: int) -> User:
        cursor = self.db.cursor()
        cursor.execute("SELECT id, name, email, created_at FROM users WHERE id = ?", (user_id,))
        row = cursor.fetchone()
        if row:
            return User(id=row[0], name=row[1], email=row[2], created_at=row[3])
        return None
    
    def find_by_email(self, email: str) -> User:
        cursor = self.db.cursor()
        cursor.execute("SELECT id, name, email, created_at FROM users WHERE email = ?", (email,))
        row = cursor.fetchone()
        if row:
            return User(id=row[0], name=row[1], email=row[2], created_at=row[3])
        return None
    
    def save(self, user: User) -> int:
        cursor = self.db.cursor()
        if user.id:
            cursor.execute(
                "UPDATE users SET name = ?, email = ? WHERE id = ?",
                (user.name, user.email, user.id)
            )
            return user.id
        else:
            cursor.execute(
                "INSERT INTO users (name, email) VALUES (?, ?)",
                (user.name, user.email)
            )
            return cursor.lastrowid
    
    def delete(self, user_id: int) -> bool:
        cursor = self.db.cursor()
        cursor.execute("DELETE FROM users WHERE id = ?", (user_id,))
        return cursor.rowcount > 0
```

#### 业务层 (Business Layer)
```python
# business/user_service.py
from data.user_model import User
from data.user_repository import UserRepository

class ValidationError(Exception):
    pass

class UserService:
    def __init__(self, user_repo: UserRepository):
        self.user_repo = user_repo
    
    def create_user(self, name: str, email: str) -> User:
        # 业务验证
        self._validate_name(name)
        self._validate_email(email)
        
        # 检查邮箱是否已存在
        existing = self.user_repo.find_by_email(email)
        if existing:
            raise ValidationError("邮箱已被注册")
        
        # 创建用户
        user = User(name=name, email=email)
        user.id = self.user_repo.save(user)
        return user
    
    def update_user(self, user_id: int, name: str = None, email: str = None) -> User:
        user = self.user_repo.find_by_id(user_id)
        if not user:
            raise ValidationError("用户不存在")
        
        if name:
            self._validate_name(name)
            user.name = name
        
        if email:
            self._validate_email(email)
            # 检查新邮箱是否被占用
            existing = self.user_repo.find_by_email(email)
            if existing and existing.id != user_id:
                raise ValidationError("邮箱已被注册")
            user.email = email
        
        self.user_repo.save(user)
        return user
    
    def delete_user(self, user_id: int) -> bool:
        user = self.user_repo.find_by_id(user_id)
        if not user:
            raise ValidationError("用户不存在")
        return self.user_repo.delete(user_id)
    
    def get_user(self, user_id: int) -> User:
        return self.user_repo.find_by_id(user_id)
    
    def _validate_name(self, name: str):
        if not name or len(name.strip()) < 2:
            raise ValidationError("用户名至少2个字符")
        if len(name) > 50:
            raise ValidationError("用户名不能超过50个字符")
    
    def _validate_email(self, email: str):
        import re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        if not re.match(pattern, email):
            raise ValidationError("邮箱格式不正确")
```

#### 表现层 (Presentation Layer)
```python
# presentation/user_view.py
from business.user_service import UserService, ValidationError

def render_user_list(users):
    html = "<h2>用户列表</h2><ul>"
    for user in users:
        html += f"<li>{user.name} ({user.email})</li>"
    html += "</ul>"
    return html

def render_user_detail(user):
    return f"""
    <div class="user-detail">
        <h2>{user.name}</h2>
        <p>邮箱: {user.email}</p>
        <p>注册时间: {user.created_at}</p>
    </div>
    """

def render_error(error):
    return f'<div class="error">{error}</div>'

def render_success(message):
    return f'<div class="success">{message}</div>'

class UserController:
    def __init__(self, user_service: UserService):
        self.user_service = user_service
    
    def list_users(self):
        # 业务逻辑由service处理
        # 这里只负责接收请求和返回响应
        return render_user_list([])
    
    def get_user(self, user_id):
        try:
            user = self.user_service.get_user(user_id)
            if not user:
                return render_error("用户不存在")
            return render_user_detail(user)
        except ValidationError as e:
            return render_error(str(e))
    
    def create_user(self, name, email):
        try:
            user = self.user_service.create_user(name, email)
            return render_success(f"用户 {user.name} 创建成功")
        except ValidationError as e:
            return render_error(str(e))
```

---

## 依赖注入示例

```python
# main.py
import sqlite3
from data.user_repository import UserRepository
from business.user_service import UserService
from presentation.user_view import UserController

# 组装应用（依赖注入）
db = sqlite3.connect('app.db')
user_repo = UserRepository(db)
user_service = UserService(user_repo)
user_controller = UserController(user_service)

# 使用
response = user_controller.create_user("张三", "zhangsan@example.com")
print(response)
```

---

## 使用Flask的MVC示例

```python
# app.py
from flask import Flask, request, jsonify
from data.user_repository import UserRepository
from business.user_service import UserService
from presentation.user_view import UserController

app = Flask(__name__)

# 依赖注入容器
class Container:
    def __init__(self):
        self.db = sqlite3.connect('app.db')
        self.user_repo = UserRepository(self.db)
        self.user_service = UserService(self.user_repo)
        self.user_controller = UserController(self.user_service)

container = Container()

@app.route('/users', methods=['GET'])
def list_users():
    return container.user_controller.list_users()

@app.route('/users/<int:user_id>', methods=['GET'])
def get_user(user_id):
    return container.user_controller.get_user(user_id)

@app.route('/users', methods=['POST'])
def create_user():
    data = request.json
    return container.user_controller.create_user(
        data['name'], 
        data['email']
    )

if __name__ == '__main__':
    app.run(debug=True)
```
